module/bdev/lvol/vbdev_lvol_rpc.c

```diff
@@ -1505,3 +1508,272 @@ rpc_bdev_lvol_shallow_copy_status(struct spdk_jsonrpc_request *request,
+	{"size", offsetof(struct rpc_bdev_lvol_get_fragmap, size), spdk_json_decode_uint64, true},
+};
+
+static void seek_hole_done_cb(struct spdk_bdev_io *bdev_io, bool success, void *cb_arg);
+
+
+static char *
+bit_array_to_base64(const struct spdk_bit_array *ba)
+{
+	uint32_t bit_count = spdk_bit_array_capacity(ba);
+	size_t byte_count = bits_to_bytes(bit_count);
+	void *bytes;
+	char *encoded;
+	size_t total_size;
+	int rc;
+
+	bytes = spdk_zmalloc(byte_count, 0, NULL, SPDK_ENV_LCORE_ID_ANY, SPDK_MALLOC_DMA);
+	if (bytes == NULL) {
+		return NULL;
+	}
+
+	for (uint32_t i = 0; i < spdk_bit_array_capacity(ba); i++) {
+		if (spdk_bit_array_get(ba, i)) {
+			// Set the bit in bytes's correct position
+			((uint8_t *)bytes)[i / 8] |= 1 << (i % 8);
+		}
+	}
+
+	total_size = spdk_base64_get_encoded_strlen(byte_count) + 1;
+	encoded = spdk_zmalloc(total_size, 0, NULL, SPDK_ENV_LCORE_ID_ANY, SPDK_MALLOC_DMA);
+	if (encoded == NULL) {
+		spdk_free(bytes);
+		return NULL;
+	}
+
+	rc = spdk_base64_encode(encoded, bytes, byte_count);
+	if (rc != 0) {
+		SPDK_ERRLOG("Failed to encode fragmap to base64\n");
+		spdk_free(bytes);
+		spdk_free(encoded);
+		return NULL;
+	}
+
+	spdk_free(bytes);
+	return encoded;
+}
+
+static void
+get_fragmap_done(struct fragmap_io *io)
+{
+	struct spdk_json_write_ctx *w  = NULL;
+	char *encoded;
+
+	encoded = bit_array_to_base64(io->fragmap);
+	if (encoded == NULL) {
+		SPDK_ERRLOG("Failed to encode fragmap to base64\n");
+		spdk_jsonrpc_send_error_response_fmt(io->request, SPDK_JSONRPC_ERROR_INTERNAL_ERROR,
+						     "Failed to ncode fragmap");
+		goto cleanup;
+	}
+
+	w = spdk_jsonrpc_begin_result(io->request);
+	spdk_json_write_object_begin(w);
+	spdk_json_write_named_uint64(w, "cluster_size", io->cluster_size);
+	spdk_json_write_named_uint64(w, "block_size", io->block_size);
+	spdk_json_write_named_uint64(w, "num_clusters", spdk_bit_array_capacity(io->fragmap));
+	spdk_json_write_named_uint64(w, "num_used_clusters", io->num_used_clusters);
+	spdk_json_write_named_string(w, "fragmap", encoded);
+
+	spdk_json_write_object_end(w);
+	spdk_jsonrpc_end_result(io->request, w);
+
+	spdk_free(encoded);
+cleanup:
+	spdk_bit_array_free(&io->fragmap);
+	spdk_put_io_channel(io->bdev_io_channel);
+	spdk_bdev_close(io->bdev_desc);
+
+	spdk_free(io);
+}
+
+static void
+seek_data_done_cb(struct spdk_bdev_io *bdev_io, bool success, void *cb_arg)
+{
+	struct fragmap_io *io = cb_arg;
+	uint64_t next_data_offset_blocks;
+	int rc;
+
+	next_data_offset_blocks = spdk_bdev_io_get_seek_offset(bdev_io);
+	spdk_bdev_free_io(bdev_io);
+
+	if (next_data_offset_blocks == UINT64_MAX) {
+		get_fragmap_done(io);
+		return;
+	}
+
+	io->current_offset = next_data_offset_blocks * io->block_size;
+	rc = spdk_bdev_seek_hole(io->bdev_desc, io->bdev_io_channel,
+				 spdk_divide_round_up(io->current_offset, io->block_size),
+				 seek_hole_done_cb, io);
+	if (rc != 0) {
+		SPDK_ERRLOG("%s\n", strerror(-rc));
```

Comment:
`spdk_bdev_seek_hole` calls the callback function only if the return value is 0, so here you must handle the end of the get_fragmap operation

Reply:
Addressed. Thank you.